gh-138813: Fix mutable default kwargs={} in multiprocessing BaseProcess and DummyProcess to use None

zeff-ir opened the pull request:
Change the default value of the 'kwargs' parameter in DummyProcess.__init__() from {} to None to avoid mutable default argument issues. This makes DummyProcess consistent with threading.Thread and threading.Timer which already use this pattern.

Found by Linux Verification Center (linuxtesting.org) with SVACE.

Reported by: Dmitrii Chuprov <cheese@altlinux.org>

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-138813 -->
* Issue: gh-138813
<!-- /gh-issue-number -->

ZeroIntensity reviewed (commented):
The core change looks good to me, but please add a [test case](https://devguide.python.org/testing/) to `test_multiprocessing`.

gpshead commented on Lib/multiprocessing/dummy/__init__.py:
```diff
@@ -33,7 +33,9 @@
 
-    def __init__(self, group=None, target=None, name=None, args=(), kwargs={}):
+    def __init__(self, group=None, target=None, name=None, args=(), kwargs=None):
+        if kwargs is None:
```
we don't need to do this. threading.Thread already accepts None and does this itself.

gpshead commented on Lib/multiprocessing/dummy/__init__.py:
could you also fix Lib/multiprocessing/process.py's BaseProcess class to have kwargs=None be the default and adjust its `self._kwargs = dict(kwargs)` line to be `= dict(kwargs) if kwargs else {}`

zeff-ir replied:
Done, pushed the changes as requested.

gpshead commented on Misc/NEWS.d/next/Library/2025-09-13-15-31-09.gh-issue-138813.1k7qT6.rst:
```diff
@@ -0,0 +1 @@
+Fix mutable default argument in `multiprocessing.dummy.DummyProcess.__init__.`
```
ReST format, use double backticks.  or :class:`multiprocessing.process.BaseProcess` reference form.

zeff-ir replied:
Thanks for pointing that out, I wasn’t aware. I’ve updated the patch to use double backticks.

zeff-ir committed:
gh-138813: Default `BaseProcess` `kwargs` to `None` (#138814)

Set `BaseProcess.__init__(..., kwargs=None)` and initialize `kwargs` with
`dict(kwargs) if kwargs else {}`. This avoids a shared mutable default and
matches threading.Thread behavior.

Co-authored-by: Dmitrii Chuprov <cheese@altlinux.org>

ZeroIntensity commented on Misc/NEWS.d/next/Library/2025-09-15-07-02-26.gh-issue-138813.yIju93.rst:
```diff
@@ -0,0 +1 @@
+multiprocessing.BaseProcess defaults ``kwargs`` to ``None`` instead of a shared dict.
```
We should use a Sphinx reference:
```suggestion
:class:`multiprocessing.BaseProcess` defaults ``kwargs`` to ``None`` instead of a shared dictionary.
```

zeff-ir replied:
Thanks, I fixed it.

zeff-ir replied:
I tried to use `:class:`, but it seems to me that `BaseProcess` is not documented as a public class, so the reference doesn’t resolve and I get a warning during the build. So I switched back to using double backticks for `BaseProcess`.

ZeroIntensity replied:
Use :class:`!multiprocessing.BaseProcess` instead.

zeff-ir replied:
Ok, I fixed it.

ZeroIntensity reviewed (commented):
Did the test case disappear?

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gpshead committed:
DummyProcess kwargs=None (which threading.Thread accepts properly)

gpshead committed:
news typo
